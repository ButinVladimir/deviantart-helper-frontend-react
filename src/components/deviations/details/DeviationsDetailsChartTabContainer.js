import { connect } from 'react-redux';
import consumeConfig from '../../shared/ConfigContext';
import DeviationsDetailsChartTab from './DeviationsDetailsChartTab';
import deviationsDetailsSetMetadataActionCreator from '../../../redux/action-creators/deviations/details/set-metadata';
import * as deviationsDetails from '../../../redux/action-creators/deviations/details/change-form-field-values';

/**
 * @description
 * Maps Redux state to props.
 *
 * @param {State} state - The Redux state.
 * @returns {Object} Props.
 */
const mapStateToProps = state => ({
  id: state.deviations.details.id,
  title: state.deviations.details.title,
  timestampBegin: state.deviations.details.timestampBegin,
  timestampEnd: state.deviations.details.timestampEnd,
  metadata: state.deviations.details.metadata,
  metadataLoading: state.deviations.details.metadataLoading,
});

/**
 * @description
 * Maps Redux dispatch to props.
 *
 * @param {Function} dispatch - The Redux dispatch.
 * @param {Object} ownProps - The own component props.
 * @returns {Object} Props.
 */
const mapDispatchToProps = (dispatch, ownProps) => ({
  timestampBeginChangeHandler:
    value => dispatch(deviationsDetails.changeTimestampBeginActionCreator(value)),
  timestampEndChangeHandler:
    value => dispatch(deviationsDetails.changeTimestampEndActionCreator(value)),
  submitHandler: () => dispatch(deviationsDetailsSetMetadataActionCreator(ownProps.config)),
});

export default consumeConfig(
  connect(mapStateToProps, mapDispatchToProps)(DeviationsDetailsChartTab),
);
