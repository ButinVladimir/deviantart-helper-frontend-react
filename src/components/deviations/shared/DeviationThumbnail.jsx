import React from 'react';
import PropTypes from 'prop-types';
import Media from 'react-bulma-components/lib/components/media';
import Tag from 'react-bulma-components/lib/components/tag';

export default function DeviationThumbnail({
  title,
  thumbnail,
  url,
  publishedTime,
  nsfw,
}) {
  const publishedTimeDate = new Date(publishedTime);

  return (
    <Media.Item position="left" className="image-container">
      <div>
        <img
          alt={title}
          src={thumbnail.src}
        />
      </div>
      <div>
        <small className="has-text-grey">
          {publishedTimeDate.toLocaleString()}
        </small>
      </div>
      <div>
        <a href={url}>Visit DA page</a>
      </div>
      {nsfw && <Tag color="danger">NSFW</Tag>}
    </Media.Item>
  );
}

DeviationThumbnail.propTypes = {
  title: PropTypes.string.isRequired,
  thumbnail: PropTypes.shape({
    src: PropTypes.string.isRequired,
    width: PropTypes.number.isRequired,
    height: PropTypes.number.isRequired,
  }).isRequired,
  url: PropTypes.string.isRequired,
  publishedTime: PropTypes.number.isRequired,
  nsfw: PropTypes.bool.isRequired,
};
