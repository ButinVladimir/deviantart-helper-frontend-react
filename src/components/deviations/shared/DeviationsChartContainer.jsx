import { connect } from 'react-redux';
import Chart from '../../chart/Chart';
import * as deviationChart from '../../../redux/action-creators/deviations/chart/change-form-field-values';
import { dataSetOptions } from '../../../consts/data-sets';

/**
 * @description
 * Maps Redux state to props.
 *
 * @param {State} state - The Redux state.
 * @returns {Object} Props.
 */
const mapStateToProps = state => ({
  dataSetTitlesMap: new Map(dataSetOptions),
  dataSet: state.deviations.chart.dataSet,
  roundPeriod: state.deviations.chart.roundPeriod,
  showTime: state.deviations.chart.showTime,
  showDifferences: state.deviations.chart.showDifferences,
});

/**
 * @description
 * Maps Redux dispatch to props.
 *
 * @param {Function} dispatch - The Redux dispatch.
 * @returns {Object} Props.
 */
const mapDispatchToProps = dispatch => ({
  dataSetChangeHandler:
    e => dispatch(deviationChart.changeDataSetActionCreator(e.target.value)),
  roundPeriodChangeHandler:
    e => dispatch(deviationChart.changeRoundPeriodActionCreator(e.target.value)),
  showTimeChangeHandler:
    e => dispatch(deviationChart.changeShowTimeActionCreator(e.target.checked)),
  showDifferencesChangeHandler:
    e => dispatch(deviationChart.changeShowDifferencesActionCreator(e.target.checked)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Chart);
