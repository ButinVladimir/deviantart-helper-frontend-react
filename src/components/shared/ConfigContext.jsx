import React, { createContext } from 'react';

const ConfigContext = createContext(null);

/**
 * @description
 * Gets config context provider.
 *
 * @returns {import('react').Provider} Provider.
 */
export const getConfigProvider = () => ConfigContext.Provider;

/**
 * @description
 * HOC to inject config.
 *
 * @param {React} Component - React component.
 * @returns {(props: Object) => React} Functional component.
 */
export default Component => props => (
  <ConfigContext.Consumer>
    {config => <Component config={config} {...props} />}
  </ConfigContext.Consumer>
);
