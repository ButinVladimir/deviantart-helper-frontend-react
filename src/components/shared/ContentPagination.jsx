import React from 'react';
import PropTypes from 'prop-types';
import Section from 'react-bulma-components/lib/components/section';
import Container from 'react-bulma-components/lib/components/container';
import Pagination from 'react-bulma-components/lib/components/pagination';

export default function ContentPagination({
  page,
  pageCount,
  showPagination,
  loadPageHandler,
}) {
  return (
    <Section>
      <Container>
        {showPagination && (
          <Pagination
            // Pagination in Bulma starts from 1 while pagination on backend start from 0.
            current={page + 1}
            total={pageCount}
            delta={3}
            onChange={loadPageHandler}
          />
        )}
      </Container>
    </Section>
  );
}

ContentPagination.propTypes = {
  page: PropTypes.number.isRequired,
  pageCount: PropTypes.number.isRequired,
  showPagination: PropTypes.bool.isRequired,
  loadPageHandler: PropTypes.func.isRequired,
};
