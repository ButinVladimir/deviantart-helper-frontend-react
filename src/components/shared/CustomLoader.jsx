import React from 'react';
import Section from 'react-bulma-components/lib/components/section';
import Container from 'react-bulma-components/lib/components/container';
import Level from 'react-bulma-components/lib/components/level';
import Loader from 'react-bulma-components/lib/components/loader';

export default function CustomLoader() {
  return (
    <Section>
      <Container>
        <Level>
          <Level.Item>
            <Loader className="custom-loader" />
          </Level.Item>
        </Level>
      </Container>
    </Section>
  );
}
