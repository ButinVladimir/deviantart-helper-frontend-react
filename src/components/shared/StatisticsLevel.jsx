import React from 'react';
import PropTypes from 'prop-types';
import Level from 'react-bulma-components/lib/components/level';
import Content from 'react-bulma-components/lib/components/content';
import Heading from 'react-bulma-components/lib/components/heading';

export default function StatisticsLevel({
  titleValuePairs,
}) {
  const items = titleValuePairs.map(pair => (
    <Level.Item key={pair[0] + pair[1]} className="has-text-centered">
      <Content>
        <div>
          <Heading size={4}>{pair[1]}</Heading>
        </div>
        <div>
          <Heading size={6} subtitle textTransform="uppercase">{pair[0]}</Heading>
        </div>
      </Content>
    </Level.Item>
  ));

  return (
    <Level breakpoint="tablet">
      {items}
    </Level>
  );
}

StatisticsLevel.propTypes = {
  titleValuePairs: PropTypes.arrayOf(PropTypes.arrayOf(PropTypes.any)).isRequired,
};
