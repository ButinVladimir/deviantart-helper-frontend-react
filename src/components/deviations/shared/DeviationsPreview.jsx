import React from 'react';
import { NavLink } from 'react-router-dom';
import PropTypes from 'prop-types';
import Media from 'react-bulma-components/lib/components/media';
import Content from 'react-bulma-components/lib/components/content';
import Heading from 'react-bulma-components/lib/components/heading';
import { Field, Control, Checkbox } from 'react-bulma-components/lib/components/form';
import DeviationThumbnail from './DeviationThumbnail';
import StatisticsLevel from '../../shared/StatisticsLevel';
import { DEVIATIONS_DETAILS, ID_PARAM } from '../../../consts/routes';

export default function DeviationsPreview({
  thumbnail,
  title,
  url,
  id,
  publishedTime,
  nsfw,
  views,
  favourites,
  comments,
  downloads,
  selected,
  toggleSelectionHandler,
}) {
  return (
    <Media>
      <DeviationThumbnail
        title={title}
        thumbnail={thumbnail}
        url={url}
        publishedTime={publishedTime}
        nsfw={nsfw}
      />
      <Media.Item>
        <Content>
          <Heading size={3}>
            <NavLink to={`${DEVIATIONS_DETAILS.replace(ID_PARAM, id)}`}>
              {title}
            </NavLink>
          </Heading>
        </Content>
        <StatisticsLevel
          titleValuePairs={[['Views', views], ['Favourites', favourites], ['Comments', comments], ['Downloads', downloads]]}
        />
      </Media.Item>
      <Media.Item position="right">
        <Field>
          <Control>
            <Checkbox checked={selected} onChange={() => toggleSelectionHandler(id)} />
          </Control>
        </Field>
      </Media.Item>
    </Media>
  );
}

DeviationsPreview.propTypes = {
  ...DeviationThumbnail.PropTypes,
  title: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  views: PropTypes.number.isRequired,
  favourites: PropTypes.number.isRequired,
  comments: PropTypes.number.isRequired,
  downloads: PropTypes.number.isRequired,
  selected: PropTypes.bool.isRequired,
  toggleSelectionHandler: PropTypes.func.isRequired,
};
