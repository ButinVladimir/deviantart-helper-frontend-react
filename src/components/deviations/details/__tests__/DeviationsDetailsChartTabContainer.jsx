import React from 'react';
import thunk from 'redux-thunk';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import config from '../../../../config';
import { getConfigProvider } from '../../../shared/ConfigContext';
import createDefaultState from '../../../../redux/states/state';
import DeviationsDetailsChartTabContainer from '../DeviationsDetailsChartTabContainer';
import { DATA_SET_DOWNLOADS } from '../../../../consts/data-sets';
import { ROUND_PERIOD_4_HOURS } from '../../../../consts/round-periods';
import deviationsDetailsLoadMetadata from '../../../../redux/action-creators/deviations/details/load-metadata';

jest.mock(
  '../../../../redux/action-creators/deviations/details/load-metadata',
  () => jest.fn(
    () => ({ type: 'SUBMIT' }),
  ),
);

describe('DeviationsDetailsChartTabContainer', () => {
  const state = {
    ...createDefaultState(),
    deviations: {
      chart: {
        dataSet: DATA_SET_DOWNLOADS,
        roundPeriod: ROUND_PERIOD_4_HOURS,
        showTime: true,
        showDifferences: true,
      },
      details: {
        id: '123-456',
        title: 'Title',
        timestampBegin: 1,
        timestampEnd: null,
        metadata: null,
        metadataLoading: false,
      },
    },
  };
  const mockStore = configureStore([thunk]);

  beforeEach(() => {
    deviationsDetailsLoadMetadata.mockClear();
  });

  it('can be rendered correctly', () => {
    const store = mockStore(state);
    const ConfigProvider = getConfigProvider();
    const wrapper = mount(
      <ConfigProvider value={config}>
        <Provider store={store}>
          <DeviationsDetailsChartTabContainer />
        </Provider>
      </ConfigProvider>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('can handle submitting', () => {
    const store = mockStore(state);
    const ConfigProvider = getConfigProvider();
    const wrapper = mount(
      <ConfigProvider value={config}>
        <Provider store={store}>
          <DeviationsDetailsChartTabContainer />
        </Provider>
      </ConfigProvider>,
    );

    wrapper.find('button#submit-button').simulate('click');

    expect(deviationsDetailsLoadMetadata).toHaveBeenCalledTimes(1);
    expect(deviationsDetailsLoadMetadata).toHaveBeenCalledWith(config);
  });
});
