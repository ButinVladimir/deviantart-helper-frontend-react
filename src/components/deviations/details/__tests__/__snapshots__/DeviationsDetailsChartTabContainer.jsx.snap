// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeviationsDetailsChartTabContainer can be rendered correctly 1`] = `
Array [
  <section
    className="section"
    style={Object {}}
  >
    <div
      className="container"
      style={Object {}}
    >
      <div
        className="field is-horizontal"
        style={Object {}}
      >
        <div
          className="field-label"
          style={Object {}}
        >
          <label
            className="label"
            htmlFor={null}
            style={Object {}}
          >
            Timestamp range beginning
          </label>
        </div>
        <div
          className="field-body"
          style={Object {}}
        >
          <div
            className="field"
            style={Object {}}
          >
            <div
              className="control"
              style={Object {}}
            >
              <div>
                <div
                  className="react-datepicker__input-container"
                >
                  <input
                    autoComplete="off"
                    className="input"
                    disabled={false}
                    id="timestamp-begin"
                    name="timestampBegin"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onClick={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    readOnly={false}
                    type="text"
                    value="01.01.1970 03:00"
                  />
                  <button
                    className="react-datepicker__close-icon"
                    onClick={[Function]}
                    tabIndex={-1}
                    type="button"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="field is-horizontal"
        style={Object {}}
      >
        <div
          className="field-label is-normal"
          style={Object {}}
        >
          <label
            className="label"
            htmlFor={null}
            style={Object {}}
          >
            Timestamp range end
          </label>
        </div>
        <div
          className="field-body"
          style={Object {}}
        >
          <div
            className="field"
            style={Object {}}
          >
            <div
              className="control"
              style={Object {}}
            >
              <div>
                <div
                  className="react-datepicker__input-container"
                >
                  <input
                    autoComplete="off"
                    className="input"
                    disabled={false}
                    id="timestamp-end"
                    name="timestampEnd"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onClick={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    readOnly={false}
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="field is-grouped"
        style={Object {}}
      >
        <div
          className="field-label is-normal"
          style={Object {}}
        />
        <div
          className="field-body"
          style={Object {}}
        >
          <div
            className="field"
            style={Object {}}
          >
            <div
              className="control"
              style={Object {}}
            >
              <button
                className="is-primary button"
                disabled={false}
                id="submit-button"
                onClick={[Function]}
                style={Object {}}
                tabIndex={0}
              >
                Submit
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </section>,
  <section
    className="section"
    style={Object {}}
  >
    <div
      className="container"
      style={Object {}}
    >
      Array [
        <div
          className="columns is-multiline"
          style={Object {}}
        >
          <div
            className="column"
            style={Object {}}
          >
            <div
              className="field"
              style={Object {}}
            >
              <label
                className="label"
                htmlFor={null}
                style={Object {}}
              >
                Data set
              </label>
              <div
                className="control"
                style={Object {}}
              >
                <div
                  className="select"
                  style={Object {}}
                >
                  <select
                    disabled={false}
                    id="data-set"
                    multiple={false}
                    name="data-set"
                    onChange={[Function]}
                    readOnly={false}
                    value={4}
                  >
                    <option
                      key="1"
                      value={1}
                    >
                      Views
                    </option>
                    <option
                      key="2"
                      value={2}
                    >
                      Favourites
                    </option>
                    <option
                      key="3"
                      value={3}
                    >
                      Comments
                    </option>
                    <option
                      key="4"
                      value={4}
                    >
                      Downloads
                    </option>
                  </select>
                </div>
              </div>
            </div>
          </div>
          <div
            className="column"
            style={Object {}}
          >
            <div
              className="field"
              style={Object {}}
            >
              <label
                className="label"
                htmlFor={null}
                style={Object {}}
              >
                Round period
              </label>
              <div
                className="control"
                style={Object {}}
              >
                <div
                  className="select"
                  style={Object {}}
                >
                  <select
                    disabled={false}
                    id="round-period"
                    multiple={false}
                    name="round-period"
                    onChange={[Function]}
                    readOnly={false}
                    value={14400000}
                  >
                    <option
                      key="1"
                      value={1}
                    >
                      No rounding
                    </option>
                    <option
                      key="60000"
                      value={60000}
                    >
                      1 minute
                    </option>
                    <option
                      key="600000"
                      value={600000}
                    >
                      10 minutes
                    </option>
                    <option
                      key="3600000"
                      value={3600000}
                    >
                      1 hour
                    </option>
                    <option
                      key="14400000"
                      value={14400000}
                    >
                      4 hours
                    </option>
                    <option
                      key="43200000"
                      value={43200000}
                    >
                      12 hours
                    </option>
                    <option
                      key="86400000"
                      value={86400000}
                    >
                      1 day
                    </option>
                    <option
                      key="604800000"
                      value={604800000}
                    >
                      7 days
                    </option>
                    <option
                      key="864000000"
                      value={864000000}
                    >
                      10 days
                    </option>
                    <option
                      key="2592000000"
                      value={2592000000}
                    >
                      30 days
                    </option>
                  </select>
                </div>
              </div>
            </div>
          </div>
          <div
            className="column"
            style={Object {}}
          >
            <div
              className="field"
              style={Object {}}
            >
              <label
                className="label"
                htmlFor={null}
                style={Object {}}
              >
                Show time
              </label>
              <div
                className="control"
                style={Object {}}
              >
                <label
                  className="checkbox"
                  disabled={false}
                  style={Object {}}
                >
                  <input
                    checked={true}
                    disabled={false}
                    id="show-time"
                    name="show-time"
                    onChange={[Function]}
                    type="checkbox"
                    value=""
                  />
                </label>
              </div>
            </div>
          </div>
          <div
            className="column is-narrow"
            style={Object {}}
          >
            <div
              className="field"
              style={Object {}}
            >
              <label
                className="label"
                htmlFor={null}
                style={Object {}}
              >
                Show differences
              </label>
              <div
                className="control"
                style={Object {}}
              >
                <label
                  className="checkbox"
                  disabled={false}
                  style={Object {}}
                >
                  <input
                    checked={true}
                    disabled={false}
                    id="show-differences"
                    name="show-differences"
                    onChange={[Function]}
                    type="checkbox"
                    value=""
                  />
                </label>
              </div>
            </div>
          </div>
        </div>,
        <div
          className="level"
          style={Object {}}
        >
          <div
            className="level-item"
            style={Object {}}
          >
            <div
              className="recharts-responsive-container"
              style={
                Object {
                  "height": 550,
                  "maxHeight": undefined,
                  "minHeight": undefined,
                  "minWidth": undefined,
                  "width": "100%",
                }
              }
            >
              <div
                style={
                  Object {
                    "display": "none",
                    "height": 0,
                    "position": "absolute",
                    "visibility": "hidden",
                    "width": 0,
                  }
                }
              />
            </div>
          </div>
        </div>,
      ]
    </div>
  </section>,
]
`;
